app/controllers/administrative-units/administrative-unit/legal-structures/new.js

```diff
@@ -0,0 +1,63 @@
+import Controller from '@ember/controller';
+import { inject as service } from '@ember/service';
+import { dropTask } from 'ember-concurrency';
+import { combineFullAddress } from 'frontend-contact-hub/models/address';
+
+export default class AdministrativeUnitsAdministrativeUnitLegalStructuresNewController extends Controller {
+  @service router;
+  @service store;
+
+  @dropTask
+  *createAssociatedStructureTask(event) {
+    event.preventDefault();
+
+    let {
+      associatedStructure,
+      address,
+      legalType,
+      registration,
+      structuredIdentifier,
+    } = this.model;
+
+    yield associatedStructure.validate();
+    yield legalType.validate();
+
+    if (associatedStructure.isValid && legalType.isValid) {
+      address.fullAddress = combineFullAddress(address);
+      yield address.save();
+
+      yield legalType.save();
+
+      yield structuredIdentifier.save();
+      registration.structuredIdentifier = structuredIdentifier;
+      yield registration.save();
+
+      associatedStructure.address = address;
+      associatedStructure.legalType = legalType;
+      associatedStructure.registration = registration;
+      yield associatedStructure.save();
+
+      this.model.administrativeUnit.associatedStructures.pushObject(
+        associatedStructure
+      );
+      yield this.model.administrativeUnit.save();
+
+      this.router.replaceWith(
+        'administrative-units.administrative-unit.legal-structures.legal-structure',
+        associatedStructure.id
+      );
+    }
+  }
+
+  reset() {
+    this.removeUnsavedRecords();
+  }
+
+  removeUnsavedRecords() {
+    this.model.associatedStructure.rollbackAttributes();
```

Comment:
It seems the new record isn't destroyed like this if it's wrapped in a changeset. I would use `.destroyRecord()` instead. That probably makes sense for all of these records here.

Reply:
You can verify by going to the "new" page, canceling and then checking the data tab in the Ember inspector. It will show unsaved records for the associated-legal-structure and legal-form-type models.

Reply:
`.destroyRecord()` gave me problem, I got "Too many recursions" error. Any suggestion?

Reply:
PS. We used `.rollbackAttributes()` in the person creation page too (in which we include the validation changeset) probably we should change there too.

Reply:
Debug postoponed.